In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

In [2]:
studAssessment = pd.read_csv('./data/studentAssessment.csv')
studAssessment

id_assessment  id_student  date_submitted  is_banked  score
0                1752       11391              18          0   78.0
1                1752       28400              22          0   70.0
2                1752       31604              17          0   72.0
3                1752       32885              26          0   69.0
4                1752       38053              19          0   79.0
...               ...         ...             ...        ...    ...
173907          37443      527538             227          0   60.0
173908          37443      534672             229          0  100.0
173909          37443      546286             215          0   80.0
173910          37443      546724             230          0  100.0
173911          37443      558486             224          0   80.0

[173912 rows x 5 columns]

In [3]:
assessment = pd.read_csv('./data/assessments.csv')
assessment

code_module code_presentation  id_assessment assessment_type   date  \
0           AAA             2013J           1752             TMA   19.0   
1           AAA             2013J           1753             TMA   54.0   
2           AAA             2013J           1754             TMA  117.0   
3           AAA             2013J           1755             TMA  166.0   
4           AAA             2013J           1756             TMA  215.0   
..          ...               ...            ...             ...    ...   
201         GGG             2014J          37443             CMA  229.0   
202         GGG             2014J          37435             TMA   61.0   
203         GGG             2014J          37436             TMA  124.0   
204         GGG             2014J          37437             TMA  173.0   
205         GGG             2014J          37444            Exam  229.0   

     weight  
0      10.0  
1      20.0  
2      20.0  
3      20.0  
4      30.0  
..      ...  
201     0.0  
202     0.0  
203     0.0  
204     0.0  
205   100.0  

[206 rows x 6 columns]

# Find number of assessment whoes weight is 0

In [4]:
print(assessment.query('weight==0')['assessment_type'].value_counts())
assessment.query('weight==0').groupby(['code_module', 'code_presentation'])['id_assessment'].count()

CMA    46
TMA    10
Name: assessment_type, dtype: int64


code_module  code_presentation
BBB          2014J                1
FFF          2013B                7
             2013J                7
             2014B                7
             2014J                7
GGG          2013J                9
             2014B                9
             2014J                9
Name: id_assessment, dtype: int64

## Lets check the number of assessment with each course

In [5]:
# find out number of assessmnet associated with each course
assessment.groupby(['code_module', 'code_presentation'])['id_assessment'].count()

code_module  code_presentation
AAA          2013J                 6
             2014J                 6
BBB          2013B                12
             2013J                12
             2014B                12
             2014J                 6
CCC          2014B                10
             2014J                10
DDD          2013B                14
             2013J                 7
             2014B                 7
             2014J                 7
EEE          2013J                 5
             2014B                 5
             2014J                 5
FFF          2013B                13
             2013J                13
             2014B                13
             2014J                13
GGG          2013J                10
             2014B                10
             2014J                10
Name: id_assessment, dtype: int64

<br>
# some useful finding

- Course BBB run at 2014J has 6 assessment out of it 1 is non-weighted
- Course FFF has 13 assessment out of it 7 assessment are non-weighted and 6 assessment are weighted
- Course GGG has 10 assessment out of it 9 assessment are non-weighted and 1 assessment is weight
<br>

In [6]:
assessment.groupby(['code_module', 'code_presentation', 'assessment_type'])['weight'].sum()

code_module  code_presentation  assessment_type
AAA          2013J              Exam               100.0
                                TMA                100.0
             2014J              Exam               100.0
                                TMA                100.0
BBB          2013B              CMA                  5.0
                                Exam               100.0
                                TMA                 95.0
             2013J              CMA                  5.0
                                Exam               100.0
                                TMA                 95.0
             2014B              CMA                  5.0
                                Exam               100.0
                                TMA                 95.0
             2014J              Exam               100.0
                                TMA                100.0
CCC          2014B              CMA                 25.0
                                Exam    

<br>
# some more insights

- Course GGG has only final exam weighted
- Course CCC has double entry of final exam
- Except course GGG, every course has 100 weights for assignmnet and 100 weights for final Exam

# Lets verify, do the final exam score exists in dataset??
<br>

In [7]:
# find all the assessmnet id of Final exam
exam = list(assessment.query('assessment_type=="Exam"')['id_assessment'])
exam

[1757,
 1763,
 14990,
 15002,
 15014,
 15025,
 24290,
 40087,
 24299,
 40088,
 25340,
 25354,
 25361,
 25368,
 30713,
 30718,
 30723,
 34872,
 34885,
 34898,
 34911,
 37424,
 37434,
 37444]

In [8]:
# find all the course whoes final exam score is included in studAssessment table
lst = studAssessment[studAssessment['id_assessment'].isin(exam)]['id_assessment'].unique()

# print that courses
assessment[assessment['id_assessment'].isin(lst)]

code_module code_presentation  id_assessment assessment_type   date  \
62          CCC             2014B          24290            Exam    NaN   
72          CCC             2014J          24299            Exam    NaN   
87          DDD             2013B          25340            Exam  240.0   
94          DDD             2013J          25354            Exam  261.0   
101         DDD             2014B          25361            Exam  241.0   
108         DDD             2014J          25368            Exam    NaN   

     weight  
62    100.0  
72    100.0  
87    100.0  
94    100.0  
101   100.0  
108   100.0

# Finding

- Only course CCC and DDD has final Exam score
- Rest of the course has no record of final Exam

# Double verify the existence of Exam score on studAssessment table 

In [9]:
exam_id = list(assessment[assessment['id_assessment'].isin(lst)]['id_assessment'])
studAssessment[studAssessment['id_assessment'].isin(exam_id)]

id_assessment  id_student  date_submitted  is_banked  score
52923          24290      558914             230          0   32.0
52924          24290      559706             234          0   78.0
52925          24290      559770             230          0   54.0
52926          24290      560114             230          0   64.0
52927          24290      560311             234          0  100.0
...              ...         ...             ...        ...    ...
95980          25368      652680             243          0   91.0
95981          25368      652732             249          0   47.0
95982          25368      652965             242          0   87.0
95983          25368      653051             242          0   27.0
95984          25368      653319             250          0   64.0

[4959 rows x 5 columns]

<br>

## Yes, there is 4959 records of Exam score

# Summary

- Only two subject (CCC and DDD) has final Exam score
- Course GGG has only non-weighted assignmnet 

# what next

- We can not use final Exam score for further analysis, it sould be removed from dataset
- For course GGG, we need to use non-weighted assignment score for further analysis
    - Approach: use average score of each student 
- For course BBB(2014J) and FFF has both graded and non-graded assessment 
    - we can drop or ignore non-graded assessment score for these courses
    
# prepare the datasets

- remove final exam score 
- remove non-weighted assessmnet of course BBB(2014J) and course FFF :: not necessary --> weight*score=0
- For course GGG, calculate the average score of each student
- For rest of the course, sum the assessment score based on (weight * score / 100)

In [10]:
# remove final exam score
print('Before remove: ', studAssessment.shape)
studAssessment = studAssessment[~studAssessment['id_assessment'].isin(exam_id)]
print('After remove: ', studAssessment.shape)

Before remove:  (173912, 5)
After remove:  (168953, 5)


In [12]:
# merge the assessmnet and studAssessment table
merge_assessment = pd.merge(assessment,studAssessment, on=['id_assessment'], how='inner')
merge_assessment

code_module code_presentation  id_assessment assessment_type   date  \
0              AAA             2013J           1752             TMA   19.0   
1              AAA             2013J           1752             TMA   19.0   
2              AAA             2013J           1752             TMA   19.0   
3              AAA             2013J           1752             TMA   19.0   
4              AAA             2013J           1752             TMA   19.0   
...            ...               ...            ...             ...    ...   
168948         GGG             2014J          37437             TMA  173.0   
168949         GGG             2014J          37437             TMA  173.0   
168950         GGG             2014J          37437             TMA  173.0   
168951         GGG             2014J          37437             TMA  173.0   
168952         GGG             2014J          37437             TMA  173.0   

        weight  id_student  date_submitted  is_banked  score  
0         10.0       11391              18          0   78.0  
1         10.0       28400              22          0   70.0  
2         10.0       31604              17          0   72.0  
3         10.0       32885              26          0   69.0  
4         10.0       38053              19          0   79.0  
...        ...         ...             ...        ...    ...  
168948     0.0      652462             172          0   60.0  
168949     0.0      652539             176          0   75.0  
168950     0.0      653157             187          0   70.0  
168951     0.0      653252             171          0   70.0  
168952     0.0      653374             172          0   50.0  

[168953 rows x 10 columns]

In [14]:
# multiply score and weight for all courses except, GGG
merge_assessment['final_marks'] = merge_assessment.apply(lambda x: x[5]*x[9]/100 if x[0]!='GGG' else x[9], axis=1)
merge_assessment

code_module code_presentation  id_assessment assessment_type   date  \
0              AAA             2013J           1752             TMA   19.0   
1              AAA             2013J           1752             TMA   19.0   
2              AAA             2013J           1752             TMA   19.0   
3              AAA             2013J           1752             TMA   19.0   
4              AAA             2013J           1752             TMA   19.0   
...            ...               ...            ...             ...    ...   
168948         GGG             2014J          37437             TMA  173.0   
168949         GGG             2014J          37437             TMA  173.0   
168950         GGG             2014J          37437             TMA  173.0   
168951         GGG             2014J          37437             TMA  173.0   
168952         GGG             2014J          37437             TMA  173.0   

        weight  id_student  date_submitted  is_banked  score  final_marks  
0         10.0       11391              18          0   78.0          7.8  
1         10.0       28400              22          0   70.0          7.0  
2         10.0       31604              17          0   72.0          7.2  
3         10.0       32885              26          0   69.0          6.9  
4         10.0       38053              19          0   79.0          7.9  
...        ...         ...             ...        ...    ...          ...  
168948     0.0      652462             172          0   60.0         60.0  
168949     0.0      652539             176          0   75.0         75.0  
168950     0.0      653157             187          0   70.0         70.0  
168951     0.0      653252             171          0   70.0         70.0  
168952     0.0      653374             172          0   50.0         50.0  

[168953 rows x 11 columns]

In [16]:
final_assessment = pd.concat([merge_assessment.query('code_module!="GGG"').
                        groupby(['code_module', 'code_presentation', 'id_student'])['final_marks'].sum(),
                        merge_assessment.query('code_module=="GGG"').
                        groupby(['code_module', 'code_presentation', 'id_student'])['final_marks'].mean()
                             ])
final_assessment

code_module  code_presentation  id_student
AAA          2013J              11391         82.400000
                                28400         65.400000
                                31604         76.300000
                                32885         55.000000
                                38053         66.900000
                                                ...    
GGG          2014J              2620947       88.888889
                                2645731       88.111111
                                2648187       76.666667
                                2679821       91.500000
                                2684003       82.857143
Name: final_marks, Length: 25839, dtype: float64

In [23]:
final_assessment = final_assessment.reset_index()
final_assessment['final_marks'] = final_assessment['final_marks'].map(lambda x: round(x,2))
final_assessment

code_module code_presentation  id_student  final_marks
0             AAA             2013J       11391        82.40
1             AAA             2013J       28400        65.40
2             AAA             2013J       31604        76.30
3             AAA             2013J       32885        55.00
4             AAA             2013J       38053        66.90
...           ...               ...         ...          ...
25834         GGG             2014J     2620947        88.89
25835         GGG             2014J     2645731        88.11
25836         GGG             2014J     2648187        76.67
25837         GGG             2014J     2679821        91.50
25838         GGG             2014J     2684003        82.86

[25839 rows x 4 columns]

# finding

- Only 25839 students have submitted assignmnet

# Lets merge final assessmnet data with studentInfo table

In [25]:
col_list = ['code_module', 'code_presentation', 'id_student','final_result']
studentinfo = pd.read_csv('./data/studentInfo.csv', usecols=col_list)
studentinfo

code_module code_presentation  id_student final_result
0             AAA             2013J       11391         Pass
1             AAA             2013J       28400         Pass
2             AAA             2013J       30268    Withdrawn
3             AAA             2013J       31604         Pass
4             AAA             2013J       32885         Pass
...           ...               ...         ...          ...
32588         GGG             2014J     2640965         Fail
32589         GGG             2014J     2645731  Distinction
32590         GGG             2014J     2648187         Pass
32591         GGG             2014J     2679821    Withdrawn
32592         GGG             2014J     2684003  Distinction

[32593 rows x 4 columns]

In [27]:
# merge the table
student_assessment = pd.merge(studentinfo, final_assessment, 
                              on=['code_module', 'code_presentation', 'id_student'],
                             how='left')
student_assessment

code_module code_presentation  id_student final_result  final_marks
0             AAA             2013J       11391         Pass        82.40
1             AAA             2013J       28400         Pass        65.40
2             AAA             2013J       30268    Withdrawn          NaN
3             AAA             2013J       31604         Pass        76.30
4             AAA             2013J       32885         Pass        55.00
...           ...               ...         ...          ...          ...
32588         GGG             2014J     2640965         Fail          NaN
32589         GGG             2014J     2645731  Distinction        88.11
32590         GGG             2014J     2648187         Pass        76.67
32591         GGG             2014J     2679821    Withdrawn        91.50
32592         GGG             2014J     2684003  Distinction        82.86

[32593 rows x 5 columns]

# remove null

- We know that only 25839 students have submitted assessmnet, therefore left merge will assign null values to all other stundet.

- Fill the null values by 0

# Save the table for further analysis

In [28]:
student_assessment.fillna(0, inplace=True)
student_assessment.to_csv('final_student_assessment.csv', index=False)